packages/core-contracts/contracts/proxy/UniversalProxyImplementation.sol

```diff
@@ -2,24 +2,73 @@
 pragma solidity ^0.8.0;
 
 import "../utils/ContractUtil.sol";
+import "../utils/TransformUtil.sol";
 
-abstract contract UniversalProxyImplementation is ContractUtil {
+abstract contract UniversalProxyImplementation is ContractUtil, TransformUtil {
     event Upgraded(address implementation);
 
     function _setImplementation(address newImplementation) internal virtual;
 
     function _getImplementation() internal view virtual returns (address);
 
+    function _setIsUpgrading(bool isUpgrading) internal virtual;
+
+    function _getIsUpgrading() internal view virtual returns (bool);
+
     function upgradeTo(address newImplementation) public virtual;
 
-    // WARNING: non-bricking protection was removed from the contract!!!
-    // Remove this comment when #214 is done
     function _upgradeTo(address newImplementation) internal {
         require(newImplementation != address(0), "Implementation is zero address");
         require(_isContract(newImplementation), "Implementation not a contract");
+        require(_canUpgradeInTheFuture(newImplementation), "Implementation is sterile");
 
         _setImplementation(newImplementation);
 
         emit Upgraded(newImplementation);
     }
+
+    function _canUpgradeInTheFuture(address newImplementation) private returns (bool) {
+        if (_getIsUpgrading()) {
```

Comment:
one of the issues of the original implementation was that in a real upgrade it will enter into a recursive loop swapping old and new implementation address and will revert at the end because of hitting the limit of function calls.

This happened since simulateUpgrade will do the swap here:
```
        _setImplementation(newImplementation);
        (bool rollbackSuccessful, ) = newImplementation.delegatecall(
            abi.encodeWithSignature("upgradeTo(address)", oldImplementation)
        );
```
and the original verification, the if condition will never result in true
```
       if (newImplementation == getImplementation()) {
            return true;
        }
```